### ISSUE TITLE
iproxy keeps running after app is stopped

### ISSUE THREAD

**Issue by knopp** [Dart-Code/Dart-Code#771]
This causes problem with subsequent launches because iproxy is already connected to observatory port on device.

If the app is run through flutter run (outside vscode), iproxy process quits after ctrl+c; However when the app is run through vscode and stopped, iproxy keeps running.

Might be relevant https://github.com/flutter/flutter/issues/4875

Flutter (Channel beta, v0.2.3, on Mac OS X 10.13.2 17C88, locale en-SK)

I can provide logs if you're unable to reproduce the problem.

**Comment by DanTup** [comment]
I haven't seen this problem on either of my MacBooks - can you provide exact steps to repro and logs of [Flutter Run](https://dartcode.org/docs/logging/#flutter-run) and [Observatory](https://dartcode.org/docs/logging/#observatory)?

Also, can you confirm what the process is you're seeing? I don't seem to have any process named `iproxy` even when the app is running (I'm not very familiar with the internals of how Flutter talks with the device, Dart Code just uses a fairly simple API from `flutter`).

**Comment by knopp** [comment]
Hi,

iproxy is part of libmobiledevice. It forwards local port to iOS device through USB cable. Which device are you trying this with? Are you deploying though USB or WiFi? I might be wrong but think iproxy is only used when deploying through USB. 

It launched by flutter run after determining the observatory port on iOS device.

I know that DartCode merely launch flutter run, but the difference is that when I terminate flutter run fom the console, it gets rid of the iproxy process as well, whereas when vscode terminates its flutter run process, iproxy keeps running.

How does vscode terminate the flutter run process? Does it send any signal, or by sending some command to stdin?

```ps x | grep iproxy```

```
69665   ??  S      0:00.06 /usr/local/bin/iproxy 8100 64126 f7e7ead1f9b5f3e2574cfbd9bc264fbcd5772a6c
```

In any case, here's the log:
```
[11:21:18 AM]: Spawning /Users/Matej/Projects/flutter/flutter/bin/flutter with args ["run","--machine","-t","lib/main.dart","-d","f7e7ead1f9b5f3e2574cfbd9bc264fbcd5772a6c","--preview-dart-2","--start-paused"]
[11:21:18 AM]: ..  in /Users/Matej/Projects/flutter/remote1
[11:21:18 AM]: ..  in {"FLUTTER_HOST":"VSCode"}
[11:21:20 AM]: <== [{"event":"daemon.showMessage","params":{"level":"warning","title":"Unable to list devices","message":"Unable to discover Android devices. Please run \"flutter doctor\" to diagnose potential issues"}}]
[11:21:20 AM]: <== [{"event":"app.start","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","deviceId":"f7e7ead1f9b5f3e2574cfbd9bc264fbcd5772a6c","directory":"/Users/Matej/Projects/flutter/remote1","supportsRestart":true}}]
[11:21:21 AM]: <== Launching lib/main.dart on Matej Knopp’s iPhone X in debug mode...
[11:21:27 AM]: <== Found saved certificate choice "iPhone Developer: Matej Knopp (QJW2VK38XZ)". To clear, use "flutter config".
[11:21:27 AM]: <== Signing iOS app for device deployment using developer identity: "iPhone Developer: Matej Knopp (QJW2VK38XZ)"
[11:21:28 AM]: <== [{"event":"app.progress","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","id":"0","progressId":null,"message":"Running Xcode clean..."}}]
[11:21:30 AM]: <== [{"event":"app.progress","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","id":"0","progressId":null,"finished":true}}]
[11:21:30 AM]: <== [{"event":"app.progress","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","id":"1","progressId":null,"message":"Starting Xcode build..."}}]
[11:21:44 AM]: <== [{"event":"app.progress","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","id":"1","progressId":null,"finished":true}}]
[11:21:44 AM]: <== Xcode build done
[11:21:46 AM]: <== [{"event":"app.progress","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","id":"2","progressId":null,"message":"Installing and launching..."}}]
[11:21:54 AM]: <== [{"event":"app.progress","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","id":"2","progressId":null,"finished":true}}]
[11:21:54 AM]: <== [{"event":"app.debugPort","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","port":8100,"wsUri":"ws://127.0.0.1:8100/ws","baseUri":"file:///private/var/mobile/Containers/Data/Application/9DEEBF11-1EE0-48B2-A608-E3CCDC03C959/tmp/remote1u5II5U/remote1/"}}]
[11:21:54 AM]: <== [{"event":"app.progress","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","id":"3","progressId":null,"message":"Syncing files to device Matej Knopp’s iPhone X..."}}]
[11:21:58 AM]: <== [{"event":"app.progress","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","id":"3","progressId":null,"finished":true}}]
[11:21:58 AM]: <== [{"event":"app.started","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4"}}]
[11:21:59 AM]: ==> [{"id":"1","method":"app.callServiceExtension","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","methodName":"ext.flutter.debugPaint","params":{"enabled":false}}}]
[11:21:59 AM]: ==> [{"id":"2","method":"app.callServiceExtension","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","methodName":"ext.flutter.showPerformanceOverlay","params":{"enabled":false}}}]
[11:21:59 AM]: ==> [{"id":"3","method":"app.callServiceExtension","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","methodName":"ext.flutter.repaintRainbow","params":{"enabled":false}}}]
[11:21:59 AM]: ==> [{"id":"4","method":"app.callServiceExtension","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","methodName":"ext.flutter.timeDilation","params":{"timeDilation":1}}}]
[11:21:59 AM]: ==> [{"id":"5","method":"app.callServiceExtension","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","methodName":"ext.flutter.debugAllowBanner","params":{"enabled":true}}}]
[11:21:59 AM]: ==> [{"id":"6","method":"app.callServiceExtension","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","methodName":"ext.flutter.debugPaintBaselinesEnabled","params":{"enabled":false}}}]
[11:21:59 AM]: ==> [{"id":"7","method":"app.callServiceExtension","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","methodName":"ext.flutter.debugPaint","params":{"enabled":false}}}]
[11:21:59 AM]: ==> [{"id":"8","method":"app.callServiceExtension","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","methodName":"ext.flutter.showPerformanceOverlay","params":{"enabled":false}}}]
[11:21:59 AM]: ==> [{"id":"9","method":"app.callServiceExtension","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","methodName":"ext.flutter.repaintRainbow","params":{"enabled":false}}}]
[11:21:59 AM]: ==> [{"id":"10","method":"app.callServiceExtension","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","methodName":"ext.flutter.timeDilation","params":{"timeDilation":1}}}]
[11:21:59 AM]: ==> [{"id":"11","method":"app.callServiceExtension","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","methodName":"ext.flutter.debugAllowBanner","params":{"enabled":true}}}]
[11:21:59 AM]: ==> [{"id":"12","method":"app.callServiceExtension","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4","methodName":"ext.flutter.debugPaintBaselinesEnabled","params":{"enabled":false}}}]
[11:21:59 AM]: <== [{"id":"2","result":{"enabled":"false","type":"_extensionType","method":"ext.flutter.showPerformanceOverlay"}}]
[11:21:59 AM]: <== [{"id":"3","result":{"enabled":"false","type":"_extensionType","method":"ext.flutter.repaintRainbow"}}]
[11:21:59 AM]: <== [{"id":"4","result":{"timeDilation":"1.0","type":"_extensionType","method":"ext.flutter.timeDilation"}}]
[11:21:59 AM]: <== [{"id":"5","result":{"enabled":"true","type":"_extensionType","method":"ext.flutter.debugAllowBanner"}}]
[11:21:59 AM]: <== [{"id":"6","result":{"enabled":"false","type":"_extensionType","method":"ext.flutter.debugPaintBaselinesEnabled"}}]
[11:21:59 AM]: <== [{"id":"7","result":{"enabled":"false","type":"_extensionType","method":"ext.flutter.debugPaint"}}]
[11:21:59 AM]: <== [{"id":"8","result":{"enabled":"false","type":"_extensionType","method":"ext.flutter.showPerformanceOverlay"}}]
[11:21:59 AM]: <== [{"id":"9","result":{"enabled":"false","type":"_extensionType","method":"ext.flutter.repaintRainbow"}}]
[11:21:59 AM]: <== [{"id":"10","result":{"timeDilation":"1.0","type":"_extensionType","method":"ext.flutter.timeDilation"}}]
[11:21:59 AM]: <== [{"id":"11","result":{"enabled":"true","type":"_extensionType","method":"ext.flutter.debugAllowBanner"}}]
[11:21:59 AM]: <== [{"id":"12","result":{"enabled":"false","type":"_extensionType","method":"ext.flutter.debugPaintBaselinesEnabled"}}]
[11:21:59 AM]: <== [{"id":"1","result":{"enabled":"false","type":"_extensionType","method":"ext.flutter.debugPaint"}}]
[11:22:13 AM]: ==> [{"id":"13","method":"app.stop","params":{"appId":"8c50fab6-7f16-4ab9-8c5d-4ac779f0c3f4"}}]
[11:22:13 AM]: <== Application finished.
[11:22:13 AM]: <== [{"id":"13","result":true}]
```

**Comment by knopp** [comment]
[flutter-observatory.txt](https://github.com/Dart-Code/Dart-Code/files/1870915/flutter-observatory.txt)

**Comment by DanTup** [comment]
> Are you deploying though USB or WiFi? I might be wrong but think iproxy is only used when deploying through USB.

I was using the iOS simulator (I don't have an iPhone). I'm not sure whether that changes things.

> How does vscode terminate the flutter run process? Does it send any signal, or by sending some command to stdin?

We call `kill` on the process. I think this may be related to https://github.com/flutter/flutter/issues/10556 that I raised some time ago. I think there may be a race - I wanted to call `app.stop` but I had issues tearing down async from VS Code.

I'll do some more digging, as IIRC I had a response about how I was supposed to handle it in the debug adapter.

**Comment by knopp** [comment]
Well, given that the iproxy process is specifically used to forward local port on computer to a device I assume it won't be there  when running on device.

**Comment by DanTup** [comment]
Based on [this](https://github.com/Dart-Code/Dart-Code/commit/d01965f318802607380677925c5a71043ada817a) and [this comment](https://github.com/flutter/flutter/issues/10556#issuecomment-307643625) I think we send stop and then immediately kill, so I think maybe Flutter isn't responding to the kill request. Maybe after a period I should send SIGKILL, though my preference would be to let it close gracefully (otherwise I don't know if it'll leave other things hanging around that could be considered bugs).

**Comment by knopp** [comment]
Well, sending sigterm from terminal terminates the process without any issues (even after vscode "attempts" to stop it).

**Comment by DanTup** [comment]
I'm marking this as blocked on https://github.com/flutter/flutter/issues/16183 as I believe the problem is that the kill signal doesn't get passed over correctly and I think we can only reliably fix it by telling Flutter directly to terminate (setting `detached: true` doesn't seem like an option on Windows if it spawns command windows).

**Comment by DanTup** [comment]
This should've been fixed in v2.13 by #907. Please shout if you still see it.

**Comment by hariprasadiit** [comment]
is this release in latest dev channel? if so, i'm still facing this issue on `0.7.3`

```
47668   ??  S      0:00.03 /usr/local/bin/iproxy 1026 52180 4517c71cd57462f4d77e237e4ead89f44ce99fa1
48851   ??  S      0:00.15 /usr/local/bin/iproxy 1027 52343 4517c71cd57462f4d77e237e4ead89f44ce99fa1
72932   ??  S      0:00.05 /usr/local/bin/iproxy 1025 55018 4517c71cd57462f4d77e237e4ead89f44ce99fa1
79673   ??  S      0:00.04 /usr/local/bin/iproxy 1024 61074 4517c71cd57462f4d77e237e4ead89f44ce99fa1
23788 s003  S+     0:00.00 grep --color=auto --exclude-dir=.bzr --exclude-dir=CVS --exclude-dir=.git--exclude-dir=.hg --exclude-dir=.svn iproxy
52281 s003  S      0:00.02 /usr/local/bin/iproxy 1028 52989 4517c71cd57462f4d77e237e4ead89f44ce99fa1
53574 s003  S      0:00.02 /usr/local/bin/iproxy 1029 53113 4517c71cd57462f4d77e237e4ead89f44ce99fa1
```

**Comment by DanTup** [comment]
@hariprasadiit Can you reproduce on demand? Could you open a new issue and [capture an log](https://dartcode.org/docs/logging/#capture-logs-command) (including Observatory and Flutter Run categories) for a session termination that you think left a process around?

Thanks!

**Comment by hariprasadiit** [comment]
@DanTup thanks for response. It seems, for me, the issue was that iProxy couldn't send the app to the iPad, when xcode was able to install the app without any problem. Restarting the iPad solved the issue. I'm not exactly sure why this happened. If i face this problem again, i'll try to reproduce this on demand and open new issue.

thanks,
